src/app/modules/startup/internal/keycard_enter_pin_state.nim

```diff
@@ -32,31 +32,34 @@ method resolveKeycardNextState*(self: KeycardEnterPinState, keycardFlowType: str
+    keycardEvent.error == RequestParamPIN:
+    controller.setRemainingAttempts(keycardEvent.pinRetries)
+    if keycardEvent.pinRetries > 0:
+      return createState(StateType.KeycardWrongPin, self.flowType, self.getBackState)
+    return createState(StateType.KeycardMaxPinRetriesReached, self.flowType, self.getBackState)
+  if keycardFlowType == ResponseTypeValueEnterPUK and
+    keycardEvent.error.len == 0:
+    if keycardEvent.pinRetries == 0 and keycardEvent.pukRetries > 0:
       return createState(StateType.KeycardMaxPinRetriesReached, self.flowType, self.getBackState)
-    if keycardFlowType == ResponseTypeValueEnterPUK and
-      keycardEvent.error.len == 0:
-      if keycardEvent.pinRetries == 0 and keycardEvent.pukRetries > 0:
-        return createState(StateType.KeycardMaxPinRetriesReached, self.flowType, self.getBackState)
-    if keycardFlowType == ResponseTypeValueSwapCard and
-      keycardEvent.error.len > 0 and
-      keycardEvent.error == RequestParamPUKRetries:
-        controller.setKeycardData(updatePredefinedKeycardData(controller.getKeycardData(), PredefinedKeycardData.MaxPUKReached, add = true))
-        return createState(StateType.KeycardMaxPukRetriesReached, self.flowType, self.getBackState)
-    if keycardFlowType == ResponseTypeValueSwapCard and
-      keycardEvent.error.len > 0 and
-      keycardEvent.error == RequestParamFreeSlots:
-        return createState(StateType.KeycardMaxPairingSlotsReached, self.flowType, self.getBackState)
-    if keycardFlowType == ResponseTypeValueKeycardFlowResult:
-      controller.setKeycardEvent(keycardEvent)
-      if not main_constants.IS_MACOS:
-        controller.setupKeycardAccount(storeToKeychain = false, recoverAccount = true)
-        return createState(StateType.ProfileFetching, self.flowType, nil)
+  if keycardFlowType == ResponseTypeValueSwapCard and
+    keycardEvent.error.len > 0 and
+    keycardEvent.error == RequestParamPUKRetries:
+      controller.setKeycardData(updatePredefinedKeycardData(controller.getKeycardData(), PredefinedKeycardData.MaxPUKReached, add = true))
+      return createState(StateType.KeycardMaxPukRetriesReached, self.flowType, self.getBackState)
+  if keycardFlowType == ResponseTypeValueSwapCard and
+    keycardEvent.error.len > 0 and
+    keycardEvent.error == RequestParamFreeSlots:
+      return createState(StateType.KeycardMaxPairingSlotsReached, self.flowType, self.getBackState)
+  if keycardFlowType == ResponseTypeValueKeycardFlowResult:
+    controller.setKeycardEvent(keycardEvent)
+    if main_constants.SUPPORTS_FINGERPRINT:
       let backState = findBackStateWithTargetedStateType(self, StateType.RecoverOldUser)
-      return createState(StateType.Biometrics, self.flowType, backState)
\ No newline at end of file
+      return createState(StateType.Biometrics, self.flowType, backState)
+    controller.setupKeycardAccount(storeToKeychain = false, recoverAccount = true)
+    return createState(StateType.ProfileFetching, self.flowType, nil)
```

Comment:
Should we have a default state in case we have an unknown type?

Reply:
I'm not sure how this works, but I can see from the other states that we don't have to

Reply:
This function will return `nil` for any keycard flow that is not handled (supported) and in that case user won't be able to run it/move to another screen.